consensus/core/src/core.rs

```diff
@@ -111,47 +155,196 @@ impl Core {
+    /// to sanity check that everything that goes into the proposal is ensured to have a timestamp < block_timestamp
+    fn ancestors_to_propose(
+        &mut self,
+        clock_round: Round,
+        block_timestamp: BlockTimestampMs,
+    ) -> Vec<BlockRef> {
+        // Now take all the ancestors up to the clock_round (excluded) and then filter only the ones with acceptable timestamp.
+        let first_include_index = self
+            .pending_ancestors
+            .iter()
+            .position(|(block_ref, timestamp)| {
+                // We assume that our system's clock can't go backwards when we perform the check here (ex due to ntp corrections)
+                assert!(*timestamp <= block_timestamp, "Violation, ancestor block timestamp {timestamp} greater than our timestamp {block_timestamp}");
+
+                block_ref.round >= clock_round
+            })
+            .unwrap_or(self.pending_ancestors.len());
+
+        let mut taken = self.pending_ancestors.split_off(first_include_index);
+        mem::swap(&mut taken, &mut self.pending_ancestors);
+
+        // Compress the references in the block. We don't want to include an ancestors that already referenced by other blocks
+        // we are about to include.
+        let mut references_in_block: HashSet<BlockRef> = HashSet::new();
+        let ancestors = taken
+            .iter()
+            .map(|(block_ref, _timestamp)| *block_ref)
+            .collect::<Vec<_>>();
+
+        for block in self.get_blocks(&ancestors).into_iter().flatten() {
+            references_in_block.extend(block.ancestors());
+        }
+
+        // explicitly add our last produced block to end result
+        let mut to_propose = vec![self.last_own_block.reference()];
+        for block_ref in ancestors.into_iter() {
+            if !references_in_block.contains(&block_ref) {
+                to_propose.push(block_ref);
+            }
+        }
+
+        assert!(!to_propose.is_empty());
+
+        to_propose
+    }
+
+    fn get_blocks(&self, _block_refs: &[BlockRef]) -> Vec<Option<VerifiedBlock>> {
+        vec![]
+    }
+
+    /// Checks whether all the leaders of the previous quorum exist.
+    /// TODO: we can leverage some additional signal here in order to more cleverly manipulate later the leader timeout
+    /// Ex if we already have one leader - the first in order - we might don't want to wait as much.
+    fn last_quorum_leaders_exist(&self) -> bool {
+        // TODO: check that we are ready to produce a new block. This will mainly check that the leaders of the previous
+        // quorum exist.
+        let quorum_round = self.threshold_clock.get_round().saturating_sub(1);
+
+        let leaders = self.leaders(quorum_round);
+        let leaders_blocks = self.blocks_at_round(quorum_round, &leaders);
+
+        // Search for all the leaders
+        leaders.iter().all(|leader| {
```

Comment:
It seems with `blocks_at_round()` that already selects blocks from `leaders`, this check will always be true?

Reply:
The logic changed now